File: src/templates/db_sqlx.ts
import { ScaffoldConfig, DbChoice } from "../types";

function sqlxTypes(db: DbChoice): { poolOptions: string; poolType: string; module: string } {
  switch (db) {
    case "sqlx-postgres":
      return { poolOptions: "PgPoolOptions", poolType: "PgPool", module: "sqlx::postgres" };
    case "sqlx-mysql":
      return { poolOptions: "MySqlPoolOptions", poolType: "MySqlPool", module: "sqlx::mysql" };
    case "sqlx-sqlite":
      return { poolOptions: "SqlitePoolOptions", poolType: "SqlitePool", module: "sqlx::sqlite" };
    default:
      return { poolOptions: "PgPoolOptions", poolType: "PgPool", module: "sqlx::postgres" };
  }
}

export function buildDbSqlxRs(config: ScaffoldConfig): string {
  const { poolOptions, poolType, module } = sqlxTypes(config.db);
  return `use ${module}::${poolOptions};

pub type Db = sqlx::${poolType};

pub async fn connect(url: &str) -> anyhow::Result<Db> {
    let pool = ${poolOptions}::new()
        .max_connections(10)
        .connect(url)
        .await?;
    Ok(pool)
}
`;
}

File: template/fragments/rust/db_sqlx.rs
use {{module}}::{{pool_options}};
use crate::config::AppConfig;

pub type Db = sqlx::{{pool_type}};

pub async fn connect(url: &str, config: &AppConfig) -> anyhow::Result<Db> {
    let pool = {{pool_options}}::new()
        .max_connections(config.db_max_connections)
        .min_connections(config.db_min_connections)
        .connect_timeout(config.db_connect_timeout())
        .idle_timeout(config.db_idle_timeout())
        .max_lifetime(config.db_max_lifetime())
        .connect(url)
        .await?;
    Ok(pool)
}
